### ISSUE TITLE
笔记 - DeepLearning - 深度学习

### ISSUE THREAD

**Issue by zengbin93** [zengbin93/blog#15]
深度学习效果惊人，做点学习笔记在这里。

学习资料的话，推荐吴恩达在2017年推出的[deeplearning.ai](https://www.deeplearning.ai/)系列课程

**Comment by zengbin93** [comment]
# 神经网络基础

**神经网络是由具有适应性的简单单元组成的广泛并行互连的网络，能够模拟生物神经系统对真实世界物体所作出的交互反应。**     -- Kohonen, 1 988

神经网络的训练过程本质上是一个参数寻优过程。

## 神经元模型

![image](https://user-images.githubusercontent.com/24988472/39430542-afebe082-4cc0-11e8-9726-eab4ee9ad161.png)
图片来自 《机器学习（周志华）》图5.1

**Comment by zengbin93** [comment]
# 2 - 激活函数

激活函数是神经元模型的重要组成部分，常用的激活函数有：1）Sigmoid函数；2）阶跃函数；3）ReLU；4）Leaky-ReLU等。

根据处理任务的不同，选取一个合适的激活函数非常关键。

## 2.1 - 阶跃函数

阶跃函数非常简单，见下图。它的缺陷是不连续、不光滑，因此，实际操作中很少使用这个作为激活函数。

![image](https://user-images.githubusercontent.com/24988472/39431752-348dd6c6-4cc4-11e8-9929-e13048ad2eb2.png)

## 2.2 - Sigmoid函数

Sigmoid函数可以把较大范围内变化的输入值挤压到（0, 1）输出值范围。此外，SIgmoid函数的导数非常容易计算，公式如下：
<img src="https://latex.codecogs.com/svg.latex?f^{'}\left(&space;x&space;\right)&space;=&space;f\left(&space;x&space;\right)\left(&space;1&space;-&space;f\left(&space;x&space;\right)&space;\right)" title="f^{'}\left( x \right) = f\left( x \right)\left( 1 - f\left( x \right) \right)" />

![image](https://user-images.githubusercontent.com/24988472/39431837-84e78248-4cc4-11e8-9f18-61ac6b533e3e.png)

**Comment by zengbin93** [comment]
# 3 - 训练方法

## 3.1 - 误差反向传播算法（Back Propagation, BP）

以最简单的双层神经网络为例，其结构图（来自《机器学习（周志华）》图5.7）如下：

<div align="center">
<img src="https://user-images.githubusercontent.com/24988472/39435865-10bc9b82-4ccf-11e8-9884-b2158fec953b.png" width="460px">
</div>

BP算法的工作流程：**先将输入示例提供给输入层神经元，然后逐层将信号前传，直到产生输出层的结果；然后计算输出层的误差，再将误差逆向传播至隐层神经元，最后根据隐层神经元的误差来对连接权和阈值进行调整。该迭代过程循环进行，直到达到某些停止条件为止。**

BP算法的训练目标是最小化训练集上的累计误差。

**Comment by zengbin93** [comment]
# 4 - 超参数

## 4.1 - learning rate

learning rate 控制着算法每一轮迭代中的更新步长，若太大则容易振荡，太小则收敛速度又会过慢。通常设置在 [0.001, 0.1] 之间

**Comment by zengbin93** [comment]
# Others

* 神经网络的训练时间少则几个小时，多则几天甚至几周，如果训练过程程序死机了，将浪费大量的时间和资源，因此**在训练过程中应当每隔一段时间就保存一次模型**

**Comment by zengbin93** [comment]
# 5 - 优化方法

* 5.1 - Attention机制

* 5.2 - Dropout机制

**Comment by zengbin93** [comment]
# 递归神经网络 Recurrent Neural Network, RNN

RNN是包含循环的网络，可以看作是同一个神经网络的多次复制，每个神经网络模块会把消息传递给下一个，因此，RNN能够使用之前一定时间内的相关信息。

<div align=center>
<img src="https://user-images.githubusercontent.com/24988472/39462761-3dc5f6f6-4d46-11e8-954f-318a936a7720.png" width="500px">

RNN结构图（from：http://colah.github.io/posts/2015-08-Understanding-LSTMs/）
</div>
